Make standard scaler compatible to Array API

AlexanderFabisch opened the pull request:
Here's my contribution from the EuroSciPy 2023 sprint. It's still work in progress and I won't have the time to continue the work before October. So if anyone else wants to take it from here, feel free to do so.

#### Reference Issues/PRs

See also #26024


#### What does this implement/fix? Explain your changes.

Make standard scaler compatible to Array API.


#### Any other comments?

Unfortunately, the current implementation breaks some unit tests of the standard scaler that are related to dtypes. That's because I wanted to make it work for torch.float16, but maybe that is not necessary and we should just support float32 and float64.

I'll also add some comments to the diff. See below.

AlexanderFabisch committed:
Array API for standard scaler

AlexanderFabisch committed:
Fix scaling

AlexanderFabisch committed:
Add unit test for standard scaler with array API

AlexanderFabisch committed:
Fix unit test

AlexanderFabisch commented on sklearn/preprocessing/_data.py:
```diff
@@ -103,7 +104,8 @@ def _handle_zeros_in_scale(scale, copy=True, constant_mask=None):
         return scale
-    elif isinstance(scale, np.ndarray):
+    else:
+    #elif isinstance(scale, np.ndarray):  # TODO instead of checking for array, check for array API compliance, how?
```
This clearly needs to be fixed: how to check whether scale is compatible to the array API?

ogrisel replied:
I think we can keep the same branching logic unchanged.

[xp.finfo](https://data-apis.org/array-api/latest/API_specification/generated/array_api.finfo.html) should still be usable to compute the constant mask.

ogrisel replied:
As far as I know the only safe way to make a copy of an array is to use `xp.asarray` as follows:

```python
    scale = xp.asarray(scale, copy=True)
```

AlexanderFabisch commented on sklearn/preprocessing/_data.py:
```diff
@@ -870,11 +872,12 @@ def partial_fit(self, X, y=None, sample_weight=None):
             X,
             accept_sparse=("csr", "csc"),
-            dtype=FLOAT_DTYPES,
+            dtype=XP_FLOAT_DTYPES(xp),
```
I guess this should be renamed to `xp_float_dtypes`?

ogrisel replied:
I think we should rather reuse `sklearn._array_api.supported_float_dtypes(xp)` here.

AlexanderFabisch commented on sklearn/preprocessing/_data.py:
```diff
@@ -888,13 +891,13 @@ def partial_fit(self, X, y=None, sample_weight=None):
+        # transform it to an array of shape (n_features,) required by
         # incr_mean_variance_axis and _incremental_variance_axis
-        dtype = np.int64 if sample_weight is None else X.dtype
+        dtype = counter_dtype(xp, X.dtype) if sample_weight is None else X.dtype
```
This was required for torch.float16 (and maybe torch.float32 as well) as the counters will be multiplied internally and will cast the internally used float16 / float32 dtype to some larger dtype, which results in an error later.

ogrisel replied:
I don't see the point of making the range of the int counter depend on the floating point precision of `X.dtype`. I think we should always use `xp.int64` to be able to support large values in `n_samples_seen_`.

AlexanderFabisch replied:
I changed it because it will have an influence on the dtype here: https://github.com/scikit-learn/scikit-learn/blob/3d9293ad18f2e7c320ea37fa0894865b9621327e/sklearn/utils/extmath.py#L1153

When there is a cast to another dtype, this will result in an error later in the code.

ogrisel replied:
We should lazily cast the integer counts to `xp.float64` whenever dividing them (or multiplying them with floating point values). For instance:

```python
    last_over_new_count = (
        xp.astype(last_sample_count, xp.float64)
       / xp.astype(new_sample_count, xp.float64)
    )
```

AlexanderFabisch commented on sklearn/preprocessing/_data.py:
```diff
@@ -944,16 +947,16 @@ def partial_fit(self, X, y=None, sample_weight=None):
             # First pass
             if not hasattr(self, "scale_"):
-                self.mean_ = 0.0
+                self.mean_ = xp.zeros(1)[0]  # TODO is there a better way?
```
How can I create a scalar value of a certain dtype? (or: Does it actually have to be a scalar value?)

ogrisel replied:
Why not keep a Python scalar here? Broadcasting Python scalars should work for all Array API implementations, no?

ogrisel replied:
If not, we would probably also need to set the `dtype` and the `device` explicitly.

EdAbati replied:
this should be solved :)

AlexanderFabisch commented on sklearn/preprocessing/_data.py:
```diff
@@ -967,7 +970,7 @@ def partial_fit(self, X, y=None, sample_weight=None):
         # if the number of samples is the same for each feature (i.e. no
         # missing values)
-        if np.ptp(self.n_samples_seen_) == 0:
+        if np.ptp(self.n_samples_seen_) == 0:  # TODO add ptp to array api utils
```
Probably the fallback will be something like `xp.max(self.n_samples_seen_) - xp.min(self.n_samples_seen_)`.

AlexanderFabisch replied:
That's certainly more elegant. :)

I guess the reason why ptp was used in the first place was because we only have to iterate over self.n_samples_seen_ once. Doing it twice could make the code slower. I just don't know whether it is significant.

ogrisel replied:
I think it's fine because this is an iteration over the features/columns that are typically small to medium size in an ML setup.

I am not even sure that `np.ptp` would do a single pass.

AlexanderFabisch commented on sklearn/preprocessing/_data.py:
```diff
@@ -976,8 +979,13 @@ def partial_fit(self, X, y=None, sample_weight=None):
             constant_mask = _is_constant_feature(
                 self.var_, self.mean_, self.n_samples_seen_
             )
+            # TODO ugly hack, torch.sqrt does not support torch.float16!
```
Yup, `torch.sqrt(x)` does not work for `torch.float16`. We could either drop support for this dtype or my solution here was to convert to float64 before calling sqrt and than converting it back to the original dtype.

ogrisel replied:
According to #12338, we should ways get `float64` `mean_` and `var_` attributes.

AlexanderFabisch commented on sklearn/preprocessing/_data.py:
```diff
@@ -976,8 +979,13 @@ def partial_fit(self, X, y=None, sample_weight=None):
                 self.var_, self.mean_, self.n_samples_seen_
             )
+            # TODO ugly hack, torch.sqrt does not support torch.float16!
+            var = xp.empty(self.var_.shape, dtype=xp.float64)
```
I guess a simple xp.astype(..., xp.float64) could replace this and the following line.

ogrisel replied:
I agree. Let's try to do this to see how the diff looks like. The temporary `var` variable should probably be renamed to `var_float64` to make the intention behind the code more explicit.

ogrisel replied:
Actually I don't understand why we would not always get `float64` variance values here, based on #12338.

AlexanderFabisch commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -207,8 +213,9 @@ def test_standard_scaler_dtype(add_sample_weight, sparse_constructor):
         assert X.dtype == X_scaled.dtype
-        assert scaler.mean_.dtype == np.float64
-        assert scaler.scale_.dtype == np.float64
+        # TODO is it a good idea to change this behavior?
```
Something to be discussed.

ogrisel replied:
Why change this? Always using `float64` is intentional as per #12338.

AlexanderFabisch replied:
OK, I got it. :)

AlexanderFabisch commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2683,3 +2690,19 @@ def test_power_transformer_constant_feature(standardize):
+    X_xp = xp.asarray(X, device=device)
+
+    scale = StandardScaler()
+    X_trans = scale.fit_transform(X_xp)
```
The test doesn't do much at the moment.

ogrisel replied:
We should reuse the standard checker as done in #26243 and #27110.

EdAbati replied:
this should be solved

AlexanderFabisch commented on sklearn/utils/_array_api.py:
```diff
@@ -10,6 +10,10 @@
 from .fixes import parse_version
 
 
+def XP_FLOAT_DTYPES(xp):
```
TODO
- [ ] docstring
- [ ] test
- [ ] better name?

ogrisel replied:
We already have `supported_float_dtypes`. I don't think we need a new function.

AlexanderFabisch commented on sklearn/utils/_array_api.py:
```diff
@@ -10,6 +10,10 @@
 
 
+def XP_FLOAT_DTYPES(xp):
+    return xp.float64, xp.float32, xp.float16
```
Should we include float16 for all backends?

ogrisel replied:
There is already `sklearn._array_api.supported_float_dtypes(xp)` that includes `float16` only when supported by the backend. I think the new `XP_FLOAT_DTYPES` is redundant.

AlexanderFabisch replied:
float16 is supported by torch, but torch.sqrt is not supported for float16. How should we handle this?

AlexanderFabisch replied:
OK, when mean and var are computed with float64, torch.sqrt should also work.

AlexanderFabisch commented on sklearn/utils/_array_api.py:
```diff
@@ -533,3 +537,11 @@ def _estimator_with_converted_arrays(estimator, converter):
     return new_estimator
+
+
+def counter_dtype(xp, dtype):  # TODO this is an ugly hack and needs to be replaced
```
I had to figure out the integer type of the same size for the standard scaler. There should be a better way... and probably also a better name for such a function.

ogrisel replied:
As commented elsewhere, I don't think we need such a helper function. The maximum number of samples in `X` is unrelated to the floating point precision of its values: it's perfectly possible to get a very large numpy array with low precision values.

AlexanderFabisch commented on sklearn/utils/extmath.py:
```diff
@@ -1030,8 +1030,10 @@ def _safe_accumulator_op(op, x, *args, **kwargs):
-        result = op(x, *args, **kwargs, dtype=np.float64)
+    from ..utils._array_api import isdtype, get_namespace
+    xp, _ = get_namespace(x)
+    if isdtype(x.dtype, "real floating", xp=xp) and x.dtype in (xp.float32, xp.float64):  # what about int, etc.?
```
Here is actually an error, I guess the second condition should be `x.dtype in (xp.float16, xp.float32)`.

ogrisel replied:
Since `x.dtype.itemsize < 8` is not an Array API compatible way to assess the precision level of a dtype, we could instead do `xp.finfo(x.dtype).bits < 64`.

EdAbati replied:
this should be solved

AlexanderFabisch commented on sklearn/utils/extmath.py:
```diff
@@ -1092,28 +1094,30 @@ def _incremental_mean_and_var(
     `utils.sparsefuncs.incr_mean_variance_axis` and
     `utils.sparsefuncs_fast.incr_mean_variance_axis0`
     """
+    from ..utils._array_api import get_namespace
```
Should be at the top of the file.

AlexanderFabisch commented on sklearn/utils/extmath.py:
```diff
@@ -1128,11 +1132,11 @@ def _incremental_mean_and_var(
             # safer because np.float64(X*W) != np.float64(X)*np.float64(W)
             correction = _safe_accumulator_op(
-                np.matmul, sample_weight, np.where(X_nan_mask, 0, temp)
+                xp.matmul, sample_weight, np.where(X_nan_mask, 0, temp)
```
Why is np.where used here?

AlexanderFabisch commented on sklearn/utils/extmath.py:
```diff
@@ -1128,11 +1132,11 @@ def _incremental_mean_and_var(
             temp **= 2
             new_unnormalized_variance = _safe_accumulator_op(
-                np.matmul, sample_weight, np.where(X_nan_mask, 0, temp)
+                xp.matmul, sample_weight, np.where(X_nan_mask, 0, temp)
```
Why is np.where used here?

ogrisel replied:
I think this is to avoid applying the `temp` correction on entries that correspond to nan values in a row of `X`.

AlexanderFabisch commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2683,3 +2690,19 @@ def test_power_transformer_constant_feature(standardize):
+def test_standard_scaler_array_api_compliance(array_namespace, device, dtype):
+    xp, device, dtype = _array_api_for_tests(array_namespace, device, dtype)
+
+    from sklearn.datasets import make_classification
```
This should either be at the top of the file or we can just use something like `random_state.randn(n_samples, n_features)`.

EdAbati commented:
Hi @AlexanderFabisch, I'm happy to continue this if it cannot wait until October. Waiting to see what the maintainers think. :)

Here are a few things I learned while working on my PR that might be helpful if you decide to keep working on it:
- update your branch with `main` to get some useful functions like `_array_api.supported_float_dtypes`
- testing the Array API compliance could be done by using a function that looks like [this](https://github.com/scikit-learn/scikit-learn/pull/26243/files#diff-7c0844db1f200a70b0665df35b84587edfd04b66f02a126c92a2b0e1806eda7aR694)
- in other places, a scalar array is created using `xp.asarray(0.0, device=device(...))`

AlexanderFabisch commented:
> I'm happy to continue this if it cannot wait until October.

Sure, I could also give you write access to my fork if needed. That way we could collaborate better.

AlexanderFabisch committed:
Use supported_float_dtypes

AlexanderFabisch committed:
Revert test

AlexanderFabisch committed:
Replace np by xp

AlexanderFabisch committed:
Remove unused import

AlexanderFabisch committed:
Remove comment

ogrisel commented on sklearn/preprocessing/_data.py:
```diff
@@ -941,8 +944,8 @@ def partial_fit(self, X, y=None, sample_weight=None):
                 # See https://github.com/scikit-learn/scikit-learn/pull/12338
-                self.mean_ = self.mean_.astype(np.float64, copy=False)
-                self.var_ = self.var_.astype(np.float64, copy=False)
+                self.mean_ = self.mean_.astype(xp.float64, copy=False)
+                self.var_ = self.var_.astype(xp.float64, copy=False)
```
There is not `.astype` method on the standard `Array`. Instead let's use the `xp.astype` function:

```suggestion
                self.mean_ = xp.astype(self.mean_, xp.float64, copy=False)
                self.var_ = xp.astype(self.var_, xp.float64, copy=False)
```

EdAbati committed:
Update sklearn/preprocessing/_data.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

EdAbati committed:
removed counter_dtype

EdAbati committed:
reverting to scalar

EdAbati committed:
reverted to np.array when matrix is sparse

EdAbati committed:
using array api for type conversion

EdAbati committed:
using xp.max/min instead of np.ptp

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

EdAbati committed:
calling _array_api.supported_float_dtypes for consistency

EdAbati committed:
using xp.finfo(x.dtype).bits < 64 in _safe_accumulator_op

EdAbati committed:
reusing standard checker

EdAbati committed:
cast std to float64

EdAbati commented:
Hi @AlexanderFabisch , thank you for sharing the fork :) 
I continued a bit, and tried to resolve some comments based on what I saw in the other PRs.

There are still a couple of TODOs:
- the "hack" regarding `sqrt` needs to be updated
- some tests are still failing because certain operations with arrays of different types are not allowed (e.g. multiplication between int64 and float64). https://github.com/scikit-learn/scikit-learn/pull/27113#discussion_r1300280986

Another thing to bear in mind is that `device='mps'` does not support `float64`. #27232 introduces something we could use

AlexanderFabisch commented:
That looks a lot better @EdAbati . Thanks for continuing this PR.

charlesjhill commented:
Howdy, is there any momentum on this PR? It'd be a nice-to-have for my work. I could be lend a hand to push this across the finish line, if necessary :)

AlexanderFabisch commented:
Hi @charlesjhill , I think nobody is currently working on it. So I guess help would be appreciated. :)

EdAbati commented:
Hi @charlesjhill, I also forgot about this one. :) so feel free to continue.
AFAIK some `array_api`-related common functions have changed in scikit learn since this PR was opened, so you might need to update some things.

charlesjhill commented:
@AlexanderFabisch, I have what I would consider a ready-to-review branch locally, but either need push-permissions on your fork, or I can open an PR on your fork, or I can open a new PR here. Let me know what you think would be easiest ~

Commit:
Support data on other devices.

Commit:
Remove "ugly hack" since torch 2.2.2 supports sqrt w/ float16.

Commit:
Add array-api support to `_check_sample_weight`

Commit:
Fixup _safe_accumulator_op and _incremental_mean_and_var to support array-api and sample weights.

Most of the changes to _incremental_mean_and_var in this commit are to satisfy array-api-strict, which has much stricter type promotion rules than actual array libraries.

Commit:
Test at float16 w/ torch for array-api tests.

Commit:
Update tags and docstrings.

Commit:
Update docs + release notes.

AlexanderFabisch commented:
@charlesjhill Please open a pull request to my branch.

Commit:
Replace "ndarray" with "array" in docstring.

Commit:
yield_namespace_device_dtype_combinations() will now return float16 in an opt-in manner.

Commit:
Use float16 for testing `check_sample_weight_array_api` too.

AlexanderFabisch commented:
I just merged @charlesjhill's pull request.

> Most of the changed files are from fast forwards to be up-to-date with the upstream scikit-learn/main.
> 
> Changes relevant to the array api work are from 7508094df707a02d1287c572ade2e319cf7e92fb...97070c10b.
> 
> This PR addresses finalizes array-api support for StandardScaler, including with sample_weights, across all tested array-api namespaces + device/dtype pairs*, including `float16`, `torch on cuda`, and the `array-api-strict` namespace.
> 
> In doing so, changes were also made to:
> - `utils.estimator_checks.{check_array_api_input, check_array_api_input_and_values}` to support sample weights during fitting
> - `utils.extmath._safe_accumulator_op` to use the maximum precision float dtype for the namespace/device pair (usually float64, but float32 on `torch on mps`). It also manually does the type promotion since torch doesn't promote in certain operations, and the array-api compatibility wrappers don't handle `float16`.
> - `utils.extmath._incremental_mean_and_var`, mostly to appease the `array_api_strict` namespace.
> - `utils.validation._check_sample_weight` to support the array api
> 
> Tests were specifically added to test array api support for `StandardScaler` when using sample weights, and to test array api support in `validation._check_sample_weight`. The oft used `yield_namespace_device_dtype_combinations` method used for testing array api conformance can now furnish `float16` in an opt-in manner. For now, this capability is only used by the tests added in this PR.
> 
> Also, w.r.t to the "ugly hack" that was used since `torch.sqrt` didn't work with float16, I couldn't replicate on my machine so I reverted that to the more straighforward np->array-api translation. See the [diff](https://github.com/AlexanderFabisch/scikit-learn/compare/7508094df707a02d1287c572ade2e319cf7e92fb...97070c10ba1a189b1ab9d173620d1406c5e10d4a#diff-9b7f4ea2b7691c59cdf3c72e68074a0a86651d23bb1dbdd233b2dc40f6d686a6L1020-L1027)
> 
> The test suite passes locally.
> 
> * I cannot test `torch on mps` since I don't have a Mac :)

Commit:
Fixup some non-conforming docstrings

Commit:
Fixup another nonconforming docstring.

charlesjhill commented:
The CI tests were reporting that some functions in `sklearn.utils._array_api` had non-conforming docstrings, but I've been unable to reproduce locally because `test_function_docstring` isn't being tested against private functions or functions in private modules (which would apply to the functions in `utils._array_api`). However, the CI tests seem to be testing against public functions in private modules (e.g. `_array_api.supported_float_dtypes`)

Is there a configuration change I need to mimic this behavior locally?

Commit:
Add tests to _incremental_mean_and_var to test list and array-api inputs.

ogrisel commented on sklearn/utils/validation.py:
```diff
@@ -1980,9 +1989,10 @@ def _check_sample_weight(
+        input is preserved; otherwise an array with the default dtype
+        is allocated.  If `dtype` is not `None` or of the "real floating" kind,
+        the output will be the maximum precision floating-point dtype for the
+        array namespace.
```
In would rather not do that ahead of time in tools for `sklearn.utils.validation` but instead delay such an upcast to the estimators code and only done if really needed.

ogrisel replied:
Instead we should use the default float dtype of the xp namespace. For PyTorch, this is float32.

But most of the time, scikit-learn estimators should call `_check_sample_weight` with an explicit `dtype=` value, typically `dtype=X.dtype`.

charlesjhill replied:
The implemented language matches the existing non-array-api behavior, which was to coerce the data type to `float64` when an explicit, non-floating-point dtype is provided. I'm not opposed to changing the behavior to only coerce to float, however. :+1:

ogrisel replied:
`float` does not exist in Array API parlance, and is a deprecated dtype for NumPy as well.

But there is indeed a default floating point dtype that is specific to each each Array API namespace:

https://data-apis.org/array-api/latest/API_specification/data_types.html#default-data-types.

As far as I know, there is no declarative way specified in the standard to discover the default floating dtype of a given namespace. But it's possible to inspect it using the following rather cheap construct: `xp.asarray(0.0).dtype`.

ogrisel commented on sklearn/utils/validation.py:
```diff
@@ -1994,16 +2004,20 @@ def _check_sample_weight(
+    xp, _ = get_namespace(X)
+
+    if dtype is not None and not isdtype(dtype, "real floating", xp=xp):
+        dtype = max_precision_float_dtype(xp, device=device(X))
```
```suggestion
        # Promote integer weights to the default float dtype of the namespace.
        dtype = xp.asarray(1.0).dtype
```

ogrisel commented on sklearn/utils/validation.py:
```diff
@@ -1980,9 +1989,10 @@ def _check_sample_weight(
-        is be allocated.  If `dtype` is not one of `float32`, `float64`,
-        `None`, the output will be of dtype `float64`.
+        input is preserved; otherwise an array with the default dtype
+        is allocated.  If `dtype` is not `None` or of the "real floating" kind,
```
```suggestion
        is allocated.  If `dtype` is not `None` and not of the "real floating" kind,
```

ogrisel commented on doc/modules/array_api.rst:
```diff
@@ -101,6 +101,7 @@ Estimators
 - :class:`preprocessing.MaxAbsScaler`
 - :class:`preprocessing.MinMaxScaler`
 - :class:`preprocessing.Normalizer`
+- :class:`preprocessing.StandardScaler`
```
I think we should add a special marker to estimators that have a device dependent dtype handling (and therefore numerical precision guarantees). For instance we could have a link such as:

```suggestion
- :class:`preprocessing.StandardScaler` (see :ref:`float64_support`)
```

that points to a new section with that label that explains that some namespaces / device combinations such as torch with the MPS device do not support `float64` operations while the estimators usually to upcast some operations to higher precision when fed with `float32` data. For devices where this is not possible, the computation stays in `float32` at the cost of potential numerical precision problems.

ogrisel replied:
Note that the `float64` upcast was introduced in #12338 without much discussion on the motivation and possible alternatives.

In particular, if we were to make the columns of `X` contiguous (maybe one column at a time to avoid a full input data copy), `numpy.(nan)sum` would have higher numerical precision via the use of pairwise summation internally (both for float32 and float64 data) as explained in https://github.com/numpy/numpy/issues/9393#issuecomment-313925335.

However I don't want to change the current behavior of `StandardScaler` for numpy by the introduction of Array API support, so I don't a good alternative to the degraded numerical precision of `StandardScaler` on namespace/device combos that do not support `float64`.

ogrisel reviewed (commented):
Thanks for updating this PR. Here is another pass of feedback.

This PR is still marked as WIP in the title and draft as a status, but I have the feeling that it is ready for review. If not please update the description (e.g. with a TODO list) to make explicit what remains to be done.

ogrisel commented:
/cc @betatim you might also want to have a look at this PR.

betatim commented on sklearn/utils/_array_api.py:
```diff
@@ -187,10 +195,49 @@ def _union1d(a, b, xp):
+    kind : str or dtype or tuple[str | dtype]
+        Data type kind.
+
+        - If ``kind`` is a dtype, then a boolean is returned if ``dtype`` is equal to
+          ``dtype``.
```
```suggestion
        - If ``kind`` is a dtype, then ``dtype`` is compared to ``kind``.
```
I've tried to make a suggestion for an improvement that explains that if you pass a dtype object for `kind`, then `dtype` is compared exactly to it. And the result is whether `kind` and `dtype` match. However I can't find a way to do it in one sentence :-/ Maybe someone else has a brain wave.

Right now we need to edit the current sentence because it says you receive a boolean if `dtype` and `dtype` are equal. Which is probably not what we are trying to say :D

Commit:
Responding to PR comments.

- Use of `_array_api.isdtype` is now kept internal.
- Validation methods don't promote willy-nilly.
- Fix bug with erroneous use of float64 without checking if torch on mps.
- Updates to use new _array_api capabilities from upstream.

Commit:
Add doc blurb about precision when using torch on mps.

charlesjhill commented on sklearn/utils/extmath.py:
```diff
@@ -1115,25 +1128,35 @@ def _incremental_mean_and_var(
+    xp, _ = get_namespace(X)
+    # Promoting int -> float is not guaranteed by the array-api, so we cast manually.
+    # (Also, last_sample_count may be a python scalar)
+    last_sample_count = xp.asarray(last_sample_count, dtype=xp.float64)
```
It'd be nice if there was a better way to mock this use-case within torch, but alas there is not. 

https://discuss.pytorch.org/t/mock-torch-device-for-unit-testing/136620 suggests using `meta` tensors, but they support only a subset of the available torch api which limits their usefulness for unit testing. E.g., `torch.unique(...)` does not work with meta tensors.

Commit:
Fix docs

Commit:
Update test to reflect change in validation behavior.

It used to be that _check_sample_weight would coerce to a maximum precision float dtype for non-floating-point data. It was changed to the default float type, though.

betatim reviewed (commented):
Another set of comments. Mostly on small doc string things and a question on "are we trying to do too much in this PR?" - I don't know the answer but it feels like maybe we are?

Thanks a lot for adding docstrings to several functions that didn't have one before!

AlexanderFabisch commented on sklearn/preprocessing/_data.py:
```diff
@@ -106,6 +106,8 @@ def _handle_zeros_in_scale(scale, copy=True, constant_mask=None):
     # scale is an array
     else:
+        # TODO instead of checking for array, check for array API compliance, how?
+        # elif isinstance(scale, np.ndarray):
```
I think I commented the line

> elif isinstance(scale, np.ndarray):

which would fail for any other array type. Instead of this check, we should have a check for array API compatibility of `scale`.

ogrisel replied:
We don't need to do any kind of type-based input validation in such private helper functions. Input validation is typically performed at the beginning of public functions or methods (e.g. by calling `_validate_data` in the `fit`, `transform` or `predict` methods of estimators).

Commit:
Remove extraneous TODO for type checking in private method.

charlesjhill commented on sklearn/preprocessing/_data.py:
```diff
@@ -751,7 +753,7 @@ class StandardScaler(OneToOneFeatureMixin, TransformerMixin, BaseEstimator):
     Attributes
     ----------
-    scale_ : ndarray of shape (n_features,) or None
+    scale_ : array of shape (n_features,) or None
```
In a word, no. I didn't go out of my way to change the docstrings on other array-api-compatible estimators (e.g., MinMaxScaler), which still refer to `ndarrays`.

ogrisel replied:
I would rather not change the public docstrings while Array API support is still in experimental mode in scikit-learn.

If we change docstring typing conventions we would need to do that consistently in all Array API PRs and we did not do that so far.

charlesjhill replied:
FWIW the recent https://github.com/scikit-learn/scikit-learn/pull/27736 updates a public docstring to refer to arrays. Also, while I can understand Array API support being experimental, I think saying "array" vs "ndarray" wouldn't confuse users who aren't using Array API support and would still signal to users that it is Array API compatible without needing to consult the online docs or estimator tags.

With that said, I'm not opposed to using `ndarray` for public docstrings for now and punting on the change for a future PR.

adrinjalali replied:
+1 on the docstring change points raised by @ogrisel

OmarManzoor replied:
I also agree with @ogrisel

charlesjhill commented on sklearn/utils/_array_api.py:
```diff
@@ -45,7 +46,9 @@ def yield_namespaces(include_numpy_namespaces=True):
 
-def yield_namespace_device_dtype_combinations(include_numpy_namespaces=True):
+def yield_namespace_device_dtype_combinations(
+    include_numpy_namespaces=True, include_float16=False
```
`_array_api.supported_float_dtypes` returns `float16` if it's in the array namespace. The docstring for that method reads:

> `float16` is not officially part of the Array API spec at the time of writing but scikit-learn estimators and functions can choose to accept it when xp.float16 is defined.
> 
> https://data-apis.org/array-api/latest/API_specification/data_types.html

I opted to ensure StandardScaler would work with `float16` inputs as that is a common use-case for me working with GPUs. I figured it could be an incremental change to other estimators as needed.

charlesjhill replied:
I implemented it as an opt-in kwarg to allow for require tests/callers to opt-in to testing against `float16`. Would a separate method/fixture be more appropriate, you think?

charlesjhill commented on sklearn/utils/_array_api.py:
```diff
@@ -232,25 +289,106 @@ def _isdtype_single(dtype, kind, *, xp):
+    Returns
+    -------
+    tuple[data-types, ...]
+        Tuple of real floating data types supported by the provided array namespace,
+        ordered from the highest precision to lowest.
```
"real floating" is used within the array api specification to refer to non-complex floating-point data types, hence why it is used here.

E.g., `is_float_but_not_complex = xp.isdtype(my_array.dtype, "real floating")`

betatim replied:
Ah ok. I had completely forgotten about complex numbers. Let's keep it like this then

charlesjhill commented on sklearn/utils/_array_api.py:
```diff
@@ -552,16 +692,48 @@ def get_namespace(*arrays, remove_none=True, remove_types=(str,), xp=None):
-def get_namespace_and_device(*array_list, remove_none=True, remove_types=(str,)):
-    """Combination into one single function of `get_namespace` and `device`."""
+def get_namespace_and_device(
+    *array_list, remove_none=True, remove_types=(str,), xp=None
```
I kept the `xp` argument around for `get_namespace_and_device` since it just delegates to `get_namespace`, but I agree that it's odd that `get_namespace` allows a passthrough... The only functionality of `get_namespace(xp=...)` would be to check if array api compatibility is enabled in the config, but it would be weird then to have an array namespace to pass in...

In short, I think removing it from the `get_namespace*` methods makes sense, or changing the behavior to verify that the provided namespace matches the namespace of the input arrays. If it is removed, callers would just have to gate the `get_namespace` call behind a `xp is None` check.

A quick perusal through the codebase suggests the `xp` kwarg is used in:
- `linear_model._ridge._solve_svd()`
- `metrics._regression._check_reg_targets()`
- `utils._array_api`
  - `_asarray_with_order`
  - `_expit`
  - `_nansum`
  - `_nanmax`
  - `_nanmin`
  - `_ravel`
  - `get_namespace_and_device`

ogrisel replied:
I think it's good not to have to write `if xp is not None` as much as possible.

So I am ok to pass the `xp` kwarg down to `get_namespace`. But we might also want to change `get_namespace` so that it actually check at all arrays passed to `get_namspace` actually stem from `xp` when it is not `None`.

WDYT?

Maybe this change is better be done in a dedicated side PR as it seems a bit unrelated to the purpose of this PR.

ogrisel replied:
Ok I re-read @charlesjhill's reply and I changed my mind again. I think it's good to keep the skip through behavior (without additional checks) when passing `xp != None` as there are many private helper methods that call `get_namespace*` and that accept an optional `xp`. Sometimes they are called by a public method that did not do the namespace inspection and rely on the private helper to do it, other times they are called by methods that already did the inspection upper in their callers ancestry.

The alternative would be to make the `xp` parameter mandatory in all private helper methods and always call `get_namespace*` in the first public API method and nowhere else.

If we decide for the latter, I think we should do this refactoring in a dedicated PR.

charlesjhill replied:
So, to summarize, the consensus is to keep `xp` as an optional kwarg for the `get_namespace*` methods, and when it is provided, to not perform any validation (i.e., current behavior)?

Commit:
Review feedback

Commit:
Add better array_api test for sample_weight support.

ogrisel reviewed (commented):
Another pass of feedback.

charlesjhill commented on sklearn/utils/extmath.py:
```diff
@@ -1050,11 +1057,18 @@ def _safe_accumulator_op(op, x, *args, **kwargs):
+        #  Normally we can rely on type promotion in `op`, but it isn't reliable for
+        #  float16 inputs.
+        target_dtype = max_precision_float_dtype(xp, device=x_device)
+        x = xp.astype(x, target_dtype, copy=False)
```
Changing the implementation to look like:

```python
def _safe_accumulator_op(op, x, *args, **kwargs):
    xp, _, x_device = get_namespace_and_device(x)
    if xp.isdtype(x.dtype, "real floating") and xp.finfo(x.dtype).bits < 64:
        kwargs["dtype"] = max_precision_float_dtype(xp, device=x_device)
    return op(x, *args, **kwargs)
```

causes test failures with torch due to `xp.matmul` not accepting a `dtype` parameter, and torch's default dtype being `float32`. `xp.matmul` does type promotion based on its arguments, but this is rather restrictive in the array api spec (e.g. no support for float16, can't promote int -> float).

This implementation fixes it for `torch.float{32,64}`, but not for `torch.float16` since the `array_api_compat` library doesn't step in to do type promotion with float16.

```python
def _safe_accumulator_op(op, x, *args, **kwargs):
    xp, _, x_device = get_namespace_and_device(x)
    if xp.isdtype(x.dtype, "real floating") and xp.finfo(x.dtype).bits < 64:
        # We need to upcast. Some ops support this natively; others don't.
        target_dtype = max_precision_float_dtype(xp, device=x_device)

        def convert_(arr):
            return xp.astype(arr, target_dtype, copy=False)

        try:
            return op(x, *args, **kwargs, dtype=target_dtype)
        except TypeError:
            # This op doesn't support a dtype kwarg, it seems. Rely on manual
            # type promotion, at the cost of memory allocations.
            x = convert_(x)
    return op(x, *args, **kwargs)
```

Manually converting arguments fixes it for the `float16` usecase (or when using other array-api non-standard dtypes):

```python
def _safe_accumulator_op(op, x, *args, **kwargs):
    xp, _, x_device = get_namespace_and_device(x)
    if xp.isdtype(x.dtype, "real floating") and xp.finfo(x.dtype).bits < 64:
        # We need to upcast. Some ops support this natively; others don't.
        target_dtype = max_precision_float_dtype(xp, device=x_device)

        def convert_(arr):
            return xp.astype(arr, target_dtype, copy=False)

        try:
            return op(x, *args, **kwargs, dtype=target_dtype)
        except TypeError:
            # This op doesn't support a dtype kwarg, it seems. Rely on manual
            # type promotion, at the cost of memory allocations.
            x = convert_(x)
            args = [
                (convert_(arg) if hasattr(arg, "dtype") else arg) for arg in args
            ]
    return op(x, *args, **kwargs)
```

I'd go with the last implementation since it allows for the more efficient code path when you can pass a dtype to the operation (like numpy and cupy), while still supporting usecases where manual type promotion is required. Let me know what you think ~

Commit:
Review feedback.

ogrisel commented on sklearn/utils/_array_api.py:
```diff
@@ -232,25 +288,106 @@ def _isdtype_single(dtype, kind, *, xp):
+    xp : module
+        Array namespace to inspect.
+
+    device : str, default=None
```
```suggestion
    device : str or device instance from xp, default=None
```

ogrisel commented on sklearn/utils/_array_api.py:
```diff
@@ -232,25 +288,106 @@ def _isdtype_single(dtype, kind, *, xp):
-        return (xp.float64, xp.float32, xp.float16)
+    # TODO: Update to use `__array_namespace__info__()` from array-api v2023.12
+    #       when/if that becomes more widespread.
+    if xp.__name__ in {"array_api_compat.torch", "torch"} and device == "mps":
```
The following is necessary when `supported_float_dtypes` or `max_precision_float_dtype` are passed device instances (e.g. from `device(X)`).

```suggestion
    if (
        xp.__name__ in {"array_api_compat.torch", "torch"}
        and getattr(device, "type", device) == "mps"
    ):
```

ogrisel reviewed (commented):
Here is another pass of feedback.

Commit:
Move changelog blurb to v1.6

Also, fixup formatting.

Commit:
Remove array-api-related guards from private methods.

Commit:
Support `device` instance in `supported_float_dtypes`

Commit:
Do device-aware dtype validation in StandardScaler.

ogrisel commented on sklearn/utils/extmath.py:
```diff
@@ -1046,12 +1053,29 @@ def _safe_accumulator_op(op, x, *args, **kwargs):
+
+        try:
+            return op(x, *args, **kwargs, dtype=target_dtype)
+        except TypeError:
```
Instead of using a `try/except TypeError`, could you please inspect the signature of `op` to check for the presence of a parameter named `"dtype"`? That would make the code more explicit I think.

betatim replied:
A reason to inspect the signature is that it is more explicit/specific. It covers us for the case where a `TypeError` is raised for a reason other than "this function doesn't have a dtype kwarg". I don't know off the top of my head what all the reasons are that a `TypeError` could be raised, so maybe this isn't an issue. But maybe the fact that I don't know off the top of my head is a reason to be more specific?

A reason to not use inspection would be that it is slow (compared to the time spent on `op`).

OmarManzoor replied:
I think let's go for inspect then if it is more clear. I agree with you on inspect being slow though.

ogrisel reviewed (commented):
Thanks for the clarification w.r.t. the lack of `dtype` kwarg for `xp.matmul` (in PyTorch in particular). This is a bit unfortunate but I don't have any better way to deal with this limitation without making the code significantly more complex, so I think we can keep this as is (with a bit more explanations in the comments, see below).

Apart from this, the discussion in https://github.com/scikit-learn/scikit-learn/pull/27113/files#r1604511376 seems to converge to not updating the docstrings w.r.t. array types as part of this PR. Could you please revert those changes?

ogrisel commented:
@AlexanderFabisch do you think you will have time addressing the last batch of review comments or would you rather like someone else to take over that PR to finalize it?

AlexanderFabisch commented:
@charlesjhill was working on this PR during the last weeks. Are you still doing this? If not, I could continue working on it again.

AlexanderFabisch committed:
Update sklearn/utils/extmath.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

ogrisel commented:
> @charlesjhill was working on this PR during the last weeks.

Sorry, I forgot that this PR had 2 active contributors.

lucyleeow commented:
Gentle ping, are @AlexanderFabisch or @charlesjhill interested in continuing this?

AlexanderFabisch commented:
I started to rebase on the current main and will compile a list of todos. Unfortunately, this PR breaks a unit test that has been introduced recently. I hope to have a clear picture of what is left to do in the beginning of next week.

AlexanderFabisch committed:
Remove duplicate test

AlexanderFabisch committed:
Fix safe accumulation

AlexanderFabisch committed:
Undo unnecessary changes to docstrings

AlexanderFabisch committed:
Undo unnecessary change to docs

AlexanderFabisch committed:
Remove unnecessary change to docstring

AlexanderFabisch committed:
Remove unused function

AlexanderFabisch committed:
Rename function convert_ to readable convert_dtype

AlexanderFabisch committed:
Sort imports

AlexanderFabisch committed:
Lint utils.extmath

AlexanderFabisch committed:
Remove duplicate imports

AlexanderFabisch committed:
Sort imports for preprocessing._data

AlexanderFabisch committed:
Run black

AlexanderFabisch committed:
Entry for "what's new"

AlexanderFabisch commented on sklearn/utils/extmath.py:
```diff
@@ -1049,12 +1056,29 @@ def _safe_accumulator_op(op, x, *args, **kwargs):
+            return op(x, *args, **kwargs, dtype=target_dtype)
+        except TypeError:
+            # This op doesn't support a dtype kwarg, it seems. Rely on manual
+            # type promotion, at the cost of memory allocations.
```
I think this is the only open discussion. Should we keep the try/except or should we inspect the signature of the function? Are there any examples of how this was done in sklearn previously?

AlexanderFabisch commented:
I rebased and cleaned up the PR a bit. I believe there is only one open discussion at the moment about using try/except vs. inspection of a function's signature in `extmath._safe_accumulator_op`. Everything else looks good. @charlesjhill What do you think?

AlexanderFabisch committed:
Black _array_api

AlexanderFabisch committed:
Fix imports with ruff

OmarManzoor committed:
Merge branch 'main' into feature/standard_scaler_array_api

OmarManzoor committed:
Update the array api support for StandardScaler

OmarManzoor committed:
Just a minor update to the changelog

OmarManzoor committed:
Some further updates

OmarManzoor committed:
Minor update

OmarManzoor committed:
Remove the float16 code

OmarManzoor committed:
Fix the pandas issue with array_api in the test test_estimator_with_set_output

betatim commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1105,7 +1126,10 @@ def check_array_api_input(
         else:
             assert attribute.shape == est_xp_param_np.shape
-            assert attribute.dtype == est_xp_param_np.dtype
+            if _check_attributes_dtype_with_array_api(
```
Do we need to check this like this? Having to maintain a list of estimators (and namespaces) inside the function seems tedious. It seems like we could check that the attribute's dtype is either equal to the dtype of the numpy based estimator (this is what we check now) or the max precision float dtype of the namespace/device (if that isn't float64).

If that is the right logic, would it be nicer to use that instead of maintaining a separate function and skip the check?

OmarManzoor replied:
I think we can certainly try that. We would need a specific check for mps generically then which should work. I separated this out because I thought not all estimators might need this kind of a logic.

betatim commented on sklearn/utils/tests/test_extmath.py:
```diff
@@ -687,18 +691,56 @@ def rng():
 @pytest.mark.parametrize("dtype", [np.float32, np.float64])
-def test_incremental_weighted_mean_and_variance_simple(rng, dtype):
+@pytest.mark.parametrize("as_list", (True, False))
+def test_incremental_weighted_mean_and_variance_simple(rng, dtype, as_list):
     mult = 10
     X = rng.rand(1000, 20).astype(dtype) * mult
     sample_weight = rng.rand(X.shape[0]) * mult
-    mean, var, _ = _incremental_mean_and_var(X, 0, 0, 0, sample_weight=sample_weight)
+    mean, var, _ = _incremental_mean_and_var(
+        X.tolist() if as_list else X, 0, 0, 0, sample_weight=sample_weight
+    )
```
Why do we need this new test case (explicitly testing list input)?

betatim commented on sklearn/utils/tests/test_extmath.py:
```diff
@@ -687,18 +691,56 @@ def rng():
     sample_weight = rng.rand(X.shape[0]) * mult
-    mean, var, _ = _incremental_mean_and_var(X, 0, 0, 0, sample_weight=sample_weight)
+    mean, var, _ = _incremental_mean_and_var(
+        X.tolist() if as_list else X, 0, 0, 0, sample_weight=sample_weight
```
If we keep this test case I'd suggest rewriting this to move the `if` outside of the function call. I had to stare at it quite a bit to parse all the commas. Now that I have seen it, it is "clear" - but I think whoever comes to this test in the future trying to understand why it is broken will appreciate it if we make it easy to parse the code. Something like
```python
X = X.tolist() if as_list else X
mean, var, _ = _incremental_mean_and_var(X, 0, 0, 0, sample_weight=sample_weight)
```
What do you think?

OmarManzoor replied:
Sure!

betatim commented:
@OmarManzoor the script in your comment uses `PolynomialFeatures` instead of `StandardScaler`. Is that on purpose?

OmarManzoor committed:
Partially address PR suggestions

OmarManzoor committed:
Address remaining suggestion

OmarManzoor committed:
Merge branch 'main' into feature/standard_scaler_array_api

OmarManzoor committed:
Some minor refactoring

OmarManzoor commented on sklearn/preprocessing/_data.py:
```diff
@@ -83,7 +85,8 @@ def _is_constant_feature(var, mean, n_samples):
     # In scikit-learn, variance is always computed using float64 accumulators.
-    eps = np.finfo(np.float64).eps
+    xp, _ = get_namespace(var, mean)
+    eps = xp.finfo(xp.float64).eps
```
We aren't using really using the mps device here. We are just using the namespace. We can't declare an array of float64 on mps.

betatim replied:
How about using `_max_precision_float_dtype` based on `var` to get the dtype that we then pass to `xp.finfo`?

OmarManzoor replied:
Yes that can be done.

lesteve commented on sklearn/utils/extmath.py:
```diff
@@ -1053,12 +1063,33 @@ def _safe_accumulator_op(op, x, *args, **kwargs):
-        result = op(x, *args, **kwargs)
-    return result
+    xp, _, x_device = get_namespace_and_device(x)
+    if xp.isdtype(x.dtype, "real floating") and xp.finfo(x.dtype).bits < 64:
```
I think it would be more explicit to compare to the max supported dtype size rather than 64 i.e. something like this?
```py
if xp.isdtype(x.dtype, "real floating"):
    max_precision_float_dtype = _max_precision_float_dtype(xp, device=x_device)
    if xp.finfo(xdtype).bits < xp.finfo(max_precision_float_dtype).bits:
        # can upcast
```

OmarManzoor committed:
Use max prec float dtype for xp.finfo

OmarManzoor committed:
Context manager check for array api because it does not support np.errstate

OmarManzoor committed:
Merge branch 'main' into feature/standard_scaler_array_api

lesteve committed:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn into feature/standard_scaler_array_api

lesteve reviewed (commented):
Some cosmetic comments that were in draft for a while. I'll try to go back to this to have a closer look in the not too far future

OmarManzoor committed:
Merge branch 'main' into feature/standard_scaler_array_api

OmarManzoor committed:
PR suggestions

OmarManzoor committed:
Merge branch 'main' into feature/standard_scaler_array_api

OmarManzoor committed:
Merge branch 'main' into feature/standard_scaler_array_api

betatim reviewed (approved):
I think this looks good.

One thing to address in a new PR is the whole story around "everything follows X" -

OmarManzoor commented on sklearn/preprocessing/_data.py:
```diff
@@ -1106,9 +1113,9 @@ def transform(self, X, copy=None):
         else:
             if self.with_mean:
-                X -= self.mean_
+                X -= xp.astype(self.mean_, X.dtype)
```
The inner computation tries to use the maximum float available and sets it accordingly. Since StandardScaler preserves the dtype we need this here as from what I remember the self.mean_ can be set according to the max float (float64) but X is float32.

betatim replied:
Thanks for figuring it out. array-pi-strict is strict :-/

OmarManzoor committed:
Merge branch 'main' into feature/standard_scaler_array_api

lesteve commented on sklearn/preprocessing/_data.py:
```diff
@@ -1050,7 +1056,7 @@ def partial_fit(self, X, y=None, sample_weight=None):
         # if the number of samples is the same for each feature (i.e. no
         # missing values)
-        if np.ptp(self.n_samples_seen_) == 0:
+        if xp.max(self.n_samples_seen_) == xp.min(self.n_samples_seen_):
```
FYI, I double-checked and `np.ptp` does nothing smart (like computing both the `min` and `max` in a single pass) so this is fine to replace it like this.

lesteve commented:
LGTM, thanks to everyone involved in this PR over the last 2 years @AlexanderFabisch @EdAbati, @ogrisel @charlesjhill @OmarManzoor @betatim!
